### PR TITLE
Fixes ''unused-argument`` false-positives with overshadowed variable in dictionary comprehension

### PR DESCRIPTION
**Pull request by hippo91**
This PR is an answer to #1731.

The ``unused-argument`` message is emitted from the ``leave_functiondef`` method of ``VariablesChecker`` class. 
In the example given by @bokhi  the wrong emission is due to the fact that the ``not_consumed`` variable holds a reference to the concerned argument whereas it should not.
The ``_to_consume`` variable of the ``VariablesChecker`` instance is not correct in this case. The origin of the problem is located inside the ``visit_name`` method.
On line 1102, there is a test to know if the name has already been consumed. When checking for the name ``key`` corresponding to the node inside the dict comprehension (``key.items()``) the answer is that it has been consumed because there is a member of ``consumed`` which name is ``key``. But this one corresponds to the variable ``key`` of the dict comprehension (`` for key, value in ...``). Due to this confusion the name ``key`` corresponding to the node ``key.items()`` is not marked as consumed and thus leads to the emission of the false-positive message.
I just added a test to avoid this confusion. 
This test is inside the ``_has_homonym_in_upper_function_scope`` method which returns True if there is another member with the same name in an upper function scope.

As i got some problems to understand this method i decided to introduce a new class named ``NamesConsumer``. The ``_to_consume`` variable is now a list of ``NamesConsumer`` instances instead of being a list of tuples with two first members being dictionaries and the third member a string. 
It allows also to have a better printing in debug sessions thanks to the ``__repr__`` method.
I hope the method to be a little bit clearer with this change.

@PCManticore, has i do not yet fully understand this complex method could you please review this PR carefully (i'm sure you always do!). I just ensured that there are no more false positive and that i didn't break the unit tests.

### CODE REVIEW

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -336,6 +337,64 @@ def _assigned_locally(name_node):
 
     }
 
+
+NamesConsumerAtomic = collections.namedtuple("NamesConsumerAtomic",
```
Comment (PCManticore) [inline pylint/checkers/variables.py]: Maybe we can rename it to ScopeConsumer since it is used mostly to hold names belonging to a scope?

**Comment by hippo91** [comment]
@PCManticore  i merged it onto master. Should i backport it to 1.8?

**Comment by PCManticore** [comment]
Sure :)

**Comment by hippo91** [comment]
@PCManticore do all bug corrections be backported to 1.8?

**Pull request merged**